### PR TITLE
Feature/#185 delete colony

### LINKED ISSUE #185: Delete Colony
Admins and Colony owners should be able to delete their colony.

This should be a command. It may be added to the townhall gui in a danger zone.

### PR DESCRIPTION
**Pull request by Kostronor**
Closes #185 

# Changes proposed in this pull request:
- A new deleteColony method in ColonyManager to cannonically delete a colony and leave no traces
- pushed getID into IColony because we need a way to pass an IColony to ColonyManager...
  - We need to rework that system, where we only use colony...
- a new delete command to delete a colony
  - The command will only delete explicit colonyids/usernames, not your own colony

Review please @Minecolonies/maintainers

### CODE REVIEW

File: src/main/java/com/minecolonies/commands/DeleteColonyCommand.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
+package com.minecolonies.commands;
+
+import com.minecolonies.colony.Colony;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/commands/DeleteColonyCommand.java]: MINOR Remove this unused import 'com.minecolonies.colony.Colony'. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/commands/DeleteColonyCommand.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+import java.util.UUID;
+
+/**
+ * List all colonies.
+ */
+public class DeleteColonyCommand extends AbstractSingleCommand
+{
+
+    public static final  String DESC                       = "delete";
+    private static final String ID_TEXT                    = "§2ID: §f";
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/commands/DeleteColonyCommand.java]: MAJOR Remove this unused "ID_TEXT" private field. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1068)

File: src/main/java/com/minecolonies/commands/DeleteColonyCommand.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    private static final String NAME_TEXT                  = "§2 Name: §f";
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/commands/DeleteColonyCommand.java]: MAJOR Remove this unused "NAME_TEXT" private field. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1068)

File: src/main/java/com/minecolonies/commands/DeleteColonyCommand.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    private static final String MAYOR_TEXT                 = "§2Mayor: §f";
+    private static final String COORDINATES_TEXT           = "§2Coordinates: §f";
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/commands/DeleteColonyCommand.java]: MAJOR Remove this unused "COORDINATES_TEXT" private field. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1068)

File: src/main/java/com/minecolonies/commands/DeleteColonyCommand.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    private static final String COORDINATES_XYZ            = "§4x=§f%s §4y=§f%s §4z=§f%s";
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/commands/DeleteColonyCommand.java]: MAJOR Remove this unused "COORDINATES_XYZ" private field. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1068)

File: src/main/java/com/minecolonies/commands/DeleteColonyCommand.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    private static final String CITIZENS                   = "§2Citizens: §f";
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/commands/DeleteColonyCommand.java]: MAJOR Remove this unused "CITIZENS" private field. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1068)

File: src/main/java/com/minecolonies/colony/ColonyManager.java [hunk 7]
```diff
@@ -87,6 +95,36 @@ public static Colony createColony(@NotNull World w, BlockPos pos, @NotNull Entit
...
+        colonies.remove(id);
+        coloniesByWorld.get(colony.getDimension()).remove(colony);
+        final Set<World> colonyWorlds = new HashSet<>();
+        for (CitizenData citizenData : colony.getCitizens().values())
+        {
+            World world = citizenData.getCitizenEntity().getEntityWorld();
+            citizenData.getCitizenEntity().onDeath(CONSOLE_DAMAGE_SOURCE);
+            colonyWorlds.add(world);
+        }
+        for (AbstractBuilding building : colony.getBuildings().values())
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/colony/ColonyManager.java]: MINOR Local variable 'building' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/colony/ColonyManager.java [hunk 8]
```diff
@@ -1,6 +1,9 @@
 package com.minecolonies.colony;
 
 import com.minecolonies.achievements.ModAchievements;
+import com.minecolonies.blocks.AbstractBlockHut;
+import com.minecolonies.blocks.BlockHutBaker;
+import com.minecolonies.blocks.ModBlocks;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/colony/ColonyManager.java]: MINOR Remove this unused import 'com.minecolonies.blocks.ModBlocks'. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/colony/ColonyManager.java [hunk 9]
```diff
@@ -87,6 +95,36 @@ public static Colony createColony(@NotNull World w, BlockPos pos, @NotNull Entit
...
+        {
+            World world = citizenData.getCitizenEntity().getEntityWorld();
+            citizenData.getCitizenEntity().onDeath(CONSOLE_DAMAGE_SOURCE);
+            colonyWorlds.add(world);
+        }
+        for (AbstractBuilding building : colony.getBuildings().values())
+        {
+            final BlockPos location = building.getLocation();
+            building.destroy();
+            for (World world : colonyWorlds)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/colony/ColonyManager.java]: MINOR Local variable 'world' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/colony/ColonyManager.java [hunk 10]
```diff
@@ -87,6 +95,36 @@ public static Colony createColony(@NotNull World w, BlockPos pos, @NotNull Entit
...
+            {
+                if(world.getBlockState(location).getBlock() instanceof AbstractBlockHut){
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/colony/ColonyManager.java]: MINOR Move this left curly brace to the beginning of next line of code. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

File: src/main/java/com/minecolonies/colony/ColonyManager.java [hunk 11]
```diff
@@ -87,6 +95,36 @@ public static Colony createColony(@NotNull World w, BlockPos pos, @NotNull Entit
...
+     */
+    public static void deleteColony(int id)
+    {
+        final Colony colony = getColony(id);
+        colonies.remove(id);
+        coloniesByWorld.get(colony.getDimension()).remove(colony);
+        final Set<World> colonyWorlds = new HashSet<>();
+        for (CitizenData citizenData : colony.getCitizens().values())
+        {
+            World world = citizenData.getCitizenEntity().getEntityWorld();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/colony/ColonyManager.java]: MINOR Local variable 'world' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 14 issues
* MAJOR 6 major
* MINOR 8 minor

Watch the comments in this conversation to review them.

File: src/main/java/com/minecolonies/colony/ColonyManager.java [hunk 12]
```diff
@@ -87,6 +93,53 @@ public static Colony createColony(@NotNull World w, BlockPos pos, @NotNull Entit
...
+            final Colony colony = getColony(id);
+            Log.getLogger().info("Deleting colony " + id);
+            colonies.remove(id);
+            coloniesByWorld.get(colony.getDimension()).remove(colony);
+            final Set<World> colonyWorlds = new HashSet<>();
+            Log.getLogger().info("Removing citizens for " + id);
+            for (CitizenData citizenData : new ArrayList<>(colony.getCitizens().values()))
+            {
+                Log.getLogger().info("Kill Citizen " + citizenData.getName());
+                World world = citizenData.getCitizenEntity().getEntityWorld();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/colony/ColonyManager.java]: MINOR Local variable 'world' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/colony/ColonyManager.java [hunk 13]
```diff
@@ -87,6 +93,53 @@ public static Colony createColony(@NotNull World w, BlockPos pos, @NotNull Entit
...
+                colonyWorlds.add(world);
+            }
+            Log.getLogger().info("Removing buildings for " + id);
+            for (AbstractBuilding building : new ArrayList<>(colony.getBuildings().values()))
+            {
+
+                final BlockPos location = building.getLocation();
+                Log.getLogger().info("Delete Building at " + location);
+                building.destroy();
+                for (World world : colonyWorlds)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/colony/ColonyManager.java]: MINOR Local variable 'world' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/colony/ColonyManager.java [hunk 14]
```diff
@@ -87,6 +93,53 @@ public static Colony createColony(@NotNull World w, BlockPos pos, @NotNull Entit
...
+    {
+        try
+        {
+            final Colony colony = getColony(id);
+            Log.getLogger().info("Deleting colony " + id);
+            colonies.remove(id);
+            coloniesByWorld.get(colony.getDimension()).remove(colony);
+            final Set<World> colonyWorlds = new HashSet<>();
+            Log.getLogger().info("Removing citizens for " + id);
+            for (CitizenData citizenData : new ArrayList<>(colony.getCitizens().values()))
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/colony/ColonyManager.java]: MINOR Local variable 'citizenData' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 4 issues
* MINOR 4 minor

Watch the comments in this conversation to review them.

File: src/main/java/com/minecolonies/colony/ColonyManager.java (hunk 12 above)
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/colony/ColonyManager.java]: MINOR Local variable 'world' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/colony/ColonyManager.java [hunk 15]
```diff
@@ -87,6 +93,53 @@ public static Colony createColony(@NotNull World w, BlockPos pos, @NotNull Entit
...
+            {
+                Log.getLogger().info("Kill Citizen " + citizenData.getName());
+                World world = citizenData.getCitizenEntity().getEntityWorld();
+                citizenData.getCitizenEntity().onDeath(CONSOLE_DAMAGE_SOURCE);
+                colonyWorlds.add(world);
+            }
+            Log.getLogger().info("Removing buildings for " + id);
+            for (AbstractBuilding building : new ArrayList<>(colony.getBuildings().values()))
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/colony/ColonyManager.java]: MINOR Local variable 'building' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/colony/ColonyManager.java (hunk 13 above)
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/colony/ColonyManager.java]: MINOR Local variable 'world' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/colony/ColonyManager.java [hunk 16]
```diff
@@ -87,6 +93,53 @@ public static Colony createColony(@NotNull World w, BlockPos pos, @NotNull Entit
...
+            final Colony colony = getColony(id);
+            Log.getLogger().info("Deleting colony " + id);
+            colonies.remove(id);
+            coloniesByWorld.get(colony.getDimension()).remove(colony);
+            final Set<World> colonyWorlds = new HashSet<>();
+            Log.getLogger().info("Removing citizens for " + id);
+            for (final CitizenData citizenData : new ArrayList<>(colony.getCitizens().values()))
+            {
+                Log.getLogger().info("Kill Citizen " + citizenData.getName());
+                World world = citizenData.getCitizenEntity().getEntityWorld();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/colony/ColonyManager.java]: MINOR Local variable 'world' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 1 issue
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: src/main/java/com/minecolonies/commands/DeleteColonyCommand.java [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+            try
+            {
+                colonyId = Integer.parseInt(args[0]);
+            }
+            catch (NumberFormatException e)
+            {
+                mayorID = getUUIDFromName(sender, args);
+            }
+        }
+
```
Comment (cltnschlosser) [inline src/main/java/com/minecolonies/commands/DeleteColonyCommand.java]: It's an edge case, but if someone's name is only numbers we could run into an issue.

Comment (Kostronor) [inline src/main/java/com/minecolonies/commands/DeleteColonyCommand.java]: The command completion would not complete him, but I think the sensible thing would be to use it as id and use it as palyername when that fails

File: src/main/java/com/minecolonies/colony/ColonyManager.java [hunk 18]
```diff
@@ -87,6 +93,53 @@ public static Colony createColony(@NotNull World w, BlockPos pos, @NotNull Entit
...
+                colonyWorlds.add(world);
+            }
+            Log.getLogger().info("Removing buildings for " + id);
+            for (final AbstractBuilding building : new ArrayList<>(colony.getBuildings().values()))
+            {
+
+                final BlockPos location = building.getLocation();
+                Log.getLogger().info("Delete Building at " + location);
+                building.destroy();
+                for (final World world : colonyWorlds)
```
Comment (Kostronor) [inline src/main/java/com/minecolonies/colony/ColonyManager.java]: Not really (at least I found nothing searching for it...). The point is that we may later have multiple world for one colony and this way we get all worlds where citizens are in.

Comment (Kostronor) [inline src/main/java/com/minecolonies/colony/ColonyManager.java]: And I check that it is a hut block before setting it, so it isn't really dangerous. Worst case is we miss a block.

Comment (cltnschlosser) [inline src/main/java/com/minecolonies/colony/ColonyManager.java]: Or delete someone else's colony blocks

Comment (Kostronor) [inline src/main/java/com/minecolonies/colony/ColonyManager.java]: I'd say the chance that a colony block in a different world is at the same place and there are colonists in that world is really small. Compared to other chances that we take, I feel we are on the safe side here

**Comment by cltnschlosser** [comment]
I think if we suppose multiple worlds, which I don't think would work with or current code, then a building should know what world it is in. Since right now we only suppose one world I think we should only delete the buildings from the colony world. If we miss buildings that's not as bad as deleting other people's colonies.

**Comment by cltnschlosser** [comment]
The Colony object contains the world

**Pull request merged**